### ISSUE TITLE
invalid parent span XXXXX; skipping clock skew adjustment

### ISSUE THREAD

**Issue by nikitaChernyshev** [jaegertracing/jaeger#3084]
**Describe the bug**

I use Spring Cloud Sleuth. I propagate headers.
My tracing span find parentSpanId, but i can't see full tree. 

**To Reproduce**
1. I deploy my app in kubernetes/istio

2. I open Jaeger dashboard and find all metrics, bull all metics have warning:
invalid parent span IDs=71726c048cad372a; skipping clock skew adjustment.
I wait 2 days and try to refresh dashboard, but i have same problem.

3. If i use io.opentracing.contrib and setup env i can see full tree. 
but i want get full tree trace with Spring cloud sleuth without io.opentracing.contrib

**Expected behavior**
i want to see full trace tree.

**Screenshots**
My ajeger dashboard:
![akipping clock skew adjustment](https://user-images.githubusercontent.com/50965855/121498941-f6cd1680-c9ed-11eb-814b-644c5d0d6da5.png)

**Version (please complete the following information):**
 - OS: Win 10, but i use google cloud shell. 
 - Jaeger version: in deployment: image: "docker.io/jaegertracing/all-in-one:1.20"
 - Deployment: **Docker**:v20.10.5 , **Kubernetes** client v1.19.7 server v1.19.9, **istio** 1.10.0

**What troubleshooting steps did you try?**
i think this is a bug on the jaeger side, or i I can't tune env :(
Mabye i should tune a bean or configure jaeger?

**my code in GitHub**
[https://github.com/nikita111100/gRPC-vs-REST](url)

**Comment by Ashmita152** [comment]
Hi @nikita111100 

I think you are hitting this bug: https://github.com/jaegertracing/jaeger/issues/2719 which is resolved in v1.22.0 jaeger release.

**Comment by nikita111100** [comment]
Now i change version to 1.22.0, but now i have same problem :(

**Comment by leRobbe** [comment]
I'm also facing this issue..it seems to have to do with parent spans being finished or w/e before execution of follow-up methods the spans get passed to for .childOf reference.

I'm facing this issue at two different parts in my code, one of which where I don't know why its running into this problem since this exact code snippet (span creation) works at a different place just fine, but with where I'm facing the issue at it's always the same span getting dropped, thus child spans "hang in the air" and span dependencies aren't shown correctly.

The 2nd place is where I'm trying to repeat a methodA (at the beginning of which I create a new span as childOf span passed to methodA) to simulate error&retry...tried both with recursively calling methodA again with current span as parameter, and when this didn't work, with creating a new object and calling its methodA. However, both ways it's always the exact same span getting dropped and childs not showing correctly in the reference tree with "warning: invalid parent span..."
When I call the methodA only once, its child methodB is referenced correctly in the dependency tree, with no warnings.
Here a simple illustration:

correct tree when calling methodB only once:
method1
  - method2
    - methodA
      - methodB (called within methodA)

wrong references (missing parent span) when trying to recursively call methodB (or creating new object with methodB and calling its methodB):
method1
  - method2
methodB, 1st invocation (-> seems at this point span of methodA called the first time gets dropped)
methodA, 2nd invocation
  - methodB (-> for some reason here the parent is referenced correctly but obviously missing its 1st parent (methodB) of 1st invocation)

what I want for the wrong scenario is:
method1
  - method2
     - methodA
           - methodB
           - methodA
                  - methodB

In the OT guides here https://opentracing.io/guides/java/spans/ they talk about scope and activate/deactivate spans, but half of the methods mentioned on this page is not available (as in, they don't exist in the OT lib I'm using).

Note: This happens both for all-in-one strategy with in-memory storage as well as production strategy with ES backend.

**Comment by Sherlock-Holo** [comment]
I got this problem too, but I am using rust with `tracing`, `tracing_opentelemetry`, `opentelemetry` and `opentelemetry_jeager` crates

**Comment by MrCroxx** [comment]
Sorry to bother, but I face the same problem, too, with rust crates `tracing`, `tracing_opentelemetry`, `opentelemetry`, and `opentelemetry_jaeger`. My jaeger version is 1.33.0. I'm not sure if its a bug from jaeger itself or rust crates.

**Comment by hugocortes** [comment]
I'm running into this issue as well in Typescript using HTTPInstrumentation. I believe this issue is due to Jaeger matching based on trace id, not span id as seen here: https://github.com/jaegertracing/jaeger/blob/7872d1b07439c3f2d316065b1fd53e885b26a66f/model/span.go#L104

In my case, I have 3 spans which should appear as:
```
> /live
  > HTTP GET
    > /ready
```

Here is `/live` info:
```json
{
  "spanID": "82a3dd7705c4527d",
  "traceID": "7ca373f7c7e0f17df9965485593768ce",
...
}
```
`HTTP GET`:
```json
{
  "spanID": "46dfdbccdb1ddb89",
  "traceID": "7ca373f7c7e0f17df9965485593768ce",
  "references": {
    "spanID": [
      "82a3dd7705c4527d"
    ],
    "traceID": [
      "7ca373f7c7e0f17df9965485593768ce"
    ],
    "refType": [
      "CHILD_OF"
    ]
  }
...
}
```
`references.spanID` and `references.traceID` equal the parents.

And here's `/ready`:
```json
{
  "spanID": "657d01f13e7ed900",
  "traceID": "f9965485593768ce",
  "references": {
    "spanID": [
      "46dfdbccdb1ddb89"
    ],
    "traceID": [
      "f9965485593768ce"
    ],
    "refType": [
      "CHILD_OF"
    ]
  }
..
}
```

In this case, traceID was truncated to 16 characters. If I update doc to be traceID: `7ca373f7c7e0f17df9965485593768ce`, then I no longer get `invalid parent span XXXX`. Error appears to be a bit misleading.

I did open an issue in JS repository: https://github.com/open-telemetry/opentelemetry-js-contrib/issues/1023 as OTTracePropgator is truncating traceID to 16 chars, however this appears to be done across various languages so I'm not quite sure whether this is a propgator bug or a Jaeger bug due to relying on TraceID rather than SpanID.

**Comment by yurishkuro** [comment]
@hugocortes the data you showed is clearly malformed, likely due to misconfiguration of the SDKs using different id length. Jaeger wouldn't even consider the 3rd span as part of the same trace - how is it supposed to know if the trace ids are different?

**Comment by goodosoft** [comment]
have same problem,wait and refresh dashboard not work.

**Comment by jkowall** [comment]
This is an Otel question since there seems to be issues with the instrumentation, as @yurishkuro pointed out. If the data is formatted correctly from instrumentation, this should not occur.

**Comment by whitehatbuds** [comment]
Is this still happening? We setup Jaeger trace on moby/buildkit using OTEL. I notice some builds will timeout, and it seems like something hung. When checking the trace, i found the exact same error message

Invalid parent SpanID=xxxx, skipping clock skew adjustment. 

It doesn't always happen but when it does, the session is stuck at "Transport is closing" and it hungs.

**Comment by OlegZavrazhin** [comment]
Still happens, any news or fixes?

**Comment by jkowall** [comment]
Please read the comments and refrain from posting useless replies to this issue. The problem is on the instrumentation and not with Jaeger. If you are having an issue please include the details of your spans and traces preferably a JSON export and open a new issue. 

This issue is closed. 

1. View the Trace: Find the trace you want to export in the Jaeger UI and click on it to view the detailed timeline.
2. Use the Download Option: Look for a "View" dropdown or similar option, usually located in the top right corner of the trace detail view.
3. Select JSON: Choose the option to view or download the trace as a JSON file.

**Comment by thapaashesh** [comment]
@Ashmita152 Just wanted to check how you resolved this issue. I’m still noticing incomplete spans in some services (spanID not in trace, clock skew adjustment). Would love to hear what worked for you.
